### PR TITLE
elixir-ls 0.2.25 (new formula)

### LINKED ISSUE #149: Homebrew formula
I think it will be great to have the ability to install elixir-ls from Homebrew so that I've created a PR with elixir-ls formula to homebrew-core.

Anyone who wants to see elixir-ls on Homebrew, please support my PR: Homebrew/homebrew-core#35918

Thanks!

### PR DESCRIPTION
**Pull request by Br1ght0ne**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
[JakeBecker/elixir-ls](https://github.com/JakeBecker/elixir-ls/)
[elixir-ls v0.2.25](https://github.com/JakeBecker/elixir-ls/releases/tag/v0.2.25)
Closes JakeBecker/elixir-ls#149.

---
This change is [Reviewable](https://reviewable.io/reviews/homebrew/homebrew-core/44105)

### CODE REVIEW

**Comment by filalex77** [comment]
@alebcay A friendly reminder :)

Would be really great to get this merged.

**Comment by chenrui333** [comment]
Thanks @filalex77 for your contribution to Homebrew! :tada: 🥇 

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

**Comment by chenrui333** [comment]
Will need to have another run for the bottle thing.

**Comment by chenrui333** [comment]
@BrewTestBot test this please

**Comment by chenrui333** [comment]
Will take care of this PR from there.

File: Formula/elixir-ls.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  desc "Language Server Protocol implementation for Elixir"
+  homepage "https://github.com/JakeBecker/elixir-ls"
+  url "https://github.com/JakeBecker/elixir-ls/archive/v0.2.25.tar.gz"
+  sha256 "48aef4a0e795627962bde7e9c5a4fd201f92ef816fc823738d6a1702ee911ee0"
+
+  depends_on "elixir"
+
+  def install
+    # Ensure Hex and Rebar3 are installed
+    system "mix", "local.hex", "--force"
```
Comment (SMillerDev) [inline Formula/elixir-ls.rb]: Will this fetch a dependency? Cause in this case it will need to be versioned.

Comment (filalex77) [reply Formula/elixir-ls.rb]: There is a way to install a specifix version of Hex.

```sh
mix hex.install 0.20.1
```

`0.20.1` is the current version at the time of writing.

Should I just use `"0.20.1"` as the argument to `system`, or should the version for Hex be specified somewhere else?

Comment (SMillerDev) [reply Formula/elixir-ls.rb]: I have no idea how mix works. But formula aren't allowed to download unversioned things. https://docs.brew.sh/Acceptable-Formulae#we-dont-like-install-scripts-that-download-unversioned-things

Comment (filalex77) [reply Formula/elixir-ls.rb]: > Note that we now allow tools like cargo, gem and pip to download specifically versioned libraries during installation.

Can we treat `mix` as the tool of this category, and `hex` as a versioned dependency?

So, those are equivalent in Elixir and Ruby:

```sh
mix hex.install 0.20.1
gem install bundler -v 2.0.1
```

Comment (SMillerDev) [reply Formula/elixir-ls.rb]: That sounds reasonable. Is there no way to have a a lockfile or something for mix?

Comment (filalex77) [reply Formula/elixir-ls.rb]: Sadly, the lockfile (`mix.lock`) only concerns the dependencies of the application. Both `hex` and `rebar3` are package managers, and `mix local.{hex,rebar}` are the primary ways to install them.

I found the URL for rebar3, but I'm having troubles with Hex. I think that installing a direct version at build time is the best option. I'll make an update.

**Comment by filalex77** [comment]
@SMillerDev Sadly, I think that if Hex is not previously installed via `mix local.hex --force`, `mix hex.install <version>` will *not* work for some reason... It asks to install it with `mix local.hex --force` first.

Honestly, I see two solution:

- Use `mix local.hex --force --if-missing`. If Hex is installed, nothing happens; if not, it installs the latest version.
- Assume Hex is available, adding a "caveat" note for users to know about `mix local.hex --force`

**Comment by SMillerDev** [comment]
If it can't support installing a specific version I am afraid that this software is not a good fit for homebrew.

**Comment by filalex77** [comment]
@SMillerDev I found a way to easily install a specific version of Hex from an archive. Can you please help me with this?

So, I have this resource block:

```ruby
resource "hex" do
  url "https://repo.hex.pm/installs/1.8.0/hex-0.20.1.ez"
  sha256 "9f11023502ad64bcc0b01f8a8edf04df89cdd2e01dda8f590e49a5447f6d8d26"
end
```

And `mix` expects a local path to a `.ez` file:

```
mix archive.install /path/to/archive.ez
```

How can I get a path for a resource, without unpacking it like an archive?

**Comment by SMillerDev** [comment]
I think you can do `mix archive.install resource("hex")`

**Comment by filalex77** [comment]
@SMillerDev Thanks, did that.

File: Formula/elixir-ls.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  depends_on "elixir"
+
+  resource "hex" do
+    url "https://repo.hex.pm/installs/1.8.0/hex-0.20.1.ez"
+    sha256 "9f11023502ad64bcc0b01f8a8edf04df89cdd2e01dda8f590e49a5447f6d8d26"
+  end
+
+  def install
+    # Ensure Hex is installed
+    system "mix", "archive.install", resource("hex")
```
Comment (MikeMcQuaid) [inline Formula/elixir-ls.rb]:
```suggestion
    system "mix", "archive.install", resource("hex").fetch
```

File: Formula/elixir-ls.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  depends_on "elixir"
+
+  resource "hex" do
+    url "https://repo.hex.pm/installs/1.8.0/hex-0.20.1.ez"
+    sha256 "9f11023502ad64bcc0b01f8a8edf04df89cdd2e01dda8f590e49a5447f6d8d26"
+  end
+
+  def install
+    # Ensure Hex is installed
```
Comment (MikeMcQuaid) [inline Formula/elixir-ls.rb]:
```suggestion
```

The above line implies this comment

**Comment by filalex77** [comment]
@MikeMcQuaid Thanks for the suggestions!

**Comment by laixintao** [comment]
Any updates on this :)

**Comment by MikeMcQuaid** [comment]
@laixintao Any will be posted in this PR.

**Comment by iMichka** [comment]
This work has started in september, and has not made any progress since then. We have tens of requests for new formulae per week, and we can't invest so much time in fixing each build failure: if the package does not behave nicely from the beginning, it feels it will be a lot of work to make it work, and also a lot of work to maintain in the future.

I'll close this for the time being, until a working version is submitted. You can of course still maintain the formula in a separate tap if needed.

**Pull request closed without merging**